### PR TITLE
Add `seed` argument for `BoTorchSampler`

### PR DESCRIPTION
**Pull request by not522**
## Motivation
Add `seed` argument to make `BoTorchSampler` reproducible.

## Description of the changes
- Add `seed` argument.
- ~~Change default candidates functions to callable classes.~~

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@contramundum53 @c-bata Could you review this PR? I think the design of this PR is controversial, so let me assign you to it.

**Comment by c-bata** [comment]
Sure :+1: Thank you for your confirmation.

**Review by contramundum53 — APPROVED** [review]
I think this design is almost good. One concern is that when a `seed` is given to the `BoTorchSampler` and not to the `candidates_func`, the sampler becomes indeterministic. This behavior might be a little counter-intuitive.
But I think this could as well be one reasonable design choice.

**Review by HideakiImamura — COMMENTED** [review]
Basically, I have several concerns.
- The `candidate` functions are public on the [document](https://optuna.readthedocs.io/en/latest/reference/integration.html). They are experimental features, so I think we can change them in this PR, it should be discussed.
- Coverage appears to have decreased relatively significantly. Should we add tests for init and seeding?
- This is not a strong opinion, but how about defining a base class for the candidate function?

**Comment by contramundum53** [comment]
I think we can create the function `qei_candidates_func`, etc., that simply returns `QeiCandidatesFunc`, etc., just to keep backwards compatibility.

**Comment by not522** [comment]
I found `torch.manual_seed` is all we need. It makes this PR simpler. PTAL.

File: optuna/integration/botorch.py [hunk 1]
```diff
@@ -400,14 +402,21 @@ def __init__(
...
         self._constraints_func = constraints_func
-        self._independent_sampler = independent_sampler or RandomSampler()
+        self._independent_sampler = independent_sampler or RandomSampler(seed=seed)
         self._n_startup_trials = n_startup_trials
 
+        if seed is not None:
+            # `torch.manual_seed` makes `BoTorchSampler` and the default candidates functions
+            # reproducible. `SobolQMCNormalSampler`'s constructor has a `seed` argument, but its
+            # behavior is deterministic when the torch's seed is fixed.
+            torch.manual_seed(seed)
```
Comment (c-bata) [inline optuna/integration/botorch.py:418]: By using `torch.manual_seed`, is there something side effect that change the behavior of user program unexpectedly?

**Review by contramundum53 — APPROVED** [review]
LGTM.

File: optuna/integration/botorch.py (hunk 1 above)
Comment (not522) [reply optuna/integration/botorch.py]: I changed to use BoTorch's `manual_seed`, which is a context manager. PTAL.

File: optuna/integration/botorch.py [hunk 2]
```diff
@@ -547,6 +559,8 @@ def sample_independent(
 
     def reseed_rng(self) -> None:
         self._independent_sampler.reseed_rng()
+        if self._seed is not None:
+            self._seed = numpy.random.RandomState().randint(numpy.iinfo(numpy.int64).max)
```
[the remark this answers is not in the document]
Comment (not522) [reply optuna/integration/botorch.py:563]: I think that overflow does not occur here. But it's a bit difficult to confirm it. `2**60` seems to be good.

Note:
`randint` returns `int`, so `self._seed += 1` is safe.
`numpy.iinfo(numpy.int64).max` is `2**63-1` and `manual_seed` can receive `[-2**63, 2**64-1]`. Then overflow does not occur here.
 https://pytorch.org/docs/stable/generated/torch.manual_seed.html

[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/integration/botorch.py]: > randint returns int, so self._seed += 1 is safe.

Oh, I missed that. You are correct :bow:

**Review by c-bata — APPROVED** [review]
LGTM!

**Pull request merged**